File: wpilibj/src/test/java/edu/wpi/first/wpilibj/PIDToleranceTest.java
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2016-2019 FIRST. All Rights Reserved.                        */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

package edu.wpi.first.wpilibj;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import edu.wpi.first.wpilibj.controller.PIDController;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

class PIDToleranceTest {
  private PIDController m_pidController;
  private static final double m_setpoint = 50.0;
  private static final double m_tolerance = 10.0;
  private static final double m_range = 200;

  private static class FakeInput {
    public double m_val;

    FakeInput() {
      m_val = 0;
    }

    public double getMeasurement() {
      return m_val;
    }
  }

  private FakeInput m_inp;

  @BeforeEach
  void setUp() {
    m_inp = new FakeInput();
    m_pidController = new PIDController(0.05, 0.0, 0.0);
    m_pidController.setInputRange(-m_range / 2, m_range / 2);
  }

  @AfterEach
  void tearDown() {
    m_pidController.close();
    m_pidController = null;
  }

  @Test
  void initialToleranceTest() {
    assertFalse(m_pidController.atSetpoint());
  }

  @Test
  void absoluteToleranceTest() {
    m_pidController.setAbsoluteTolerance(m_tolerance);
    m_pidController.setSetpoint(m_setpoint);
    for (int i = 0; i < 50; i++) {
      m_pidController.calculate(m_inp.getMeasurement());
    }
    assertFalse(m_pidController.atSetpoint(),
        "Error was in tolerance when it should not have been. Error was " + m_pidController
            .getError());
    m_inp.m_val = m_setpoint + m_tolerance / 2;
    for (int i = 0; i < 50; i++) {
      m_pidController.calculate(m_inp.getMeasurement());
    }
    assertTrue(m_pidController.atSetpoint(),
        "Error was not in tolerance when it should have been. Error was " + m_pidController
            .getError());
    m_inp.m_val = m_setpoint + 10 * m_tolerance;
    for (int i = 0; i < 50; i++) {
      m_pidController.calculate(m_inp.getMeasurement());
    }
    assertFalse(m_pidController.atSetpoint(),
        "Error was in tolerance when it should not have been. Error was " + m_pidController
            .getError());
  }

  @Test
  void percentToleranceTest() {
    m_pidController.setPercentTolerance(m_tolerance);
    m_pidController.setSetpoint(m_setpoint);
    for (int i = 0; i < 50; i++) {
      m_pidController.calculate(m_inp.getMeasurement());
    }
    assertFalse(m_pidController.atSetpoint(),
        "Error was in tolerance when it should not have been. Error was " + m_pidController
            .getError());
    //half of percent tolerance away from setPoint
    m_inp.m_val = m_setpoint + m_tolerance / 200 * m_range;
    for (int i = 0; i < 50; i++) {
      m_pidController.calculate(m_inp.getMeasurement());
    }
    assertTrue(m_pidController.atSetpoint(),
        "Error was not in tolerance when it should have been. Error was " + m_pidController
            .getError());
    //double percent tolerance away from setPoint
    m_inp.m_val = m_setpoint + m_tolerance / 50 * m_range;
    for (int i = 0; i < 50; i++) {
      m_pidController.calculate(m_inp.getMeasurement());
    }
    assertFalse(m_pidController.atSetpoint(),
        "Error was in tolerance when it should not have been. Error was " + m_pidController
            .getError());
  }
}

File: wpilibc/src/test/native/cpp/PIDToleranceTest.cpp
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2014-2019 FIRST. All Rights Reserved.                        */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

#include "frc/Timer.h"
#include "frc/controller/PIDController.h"
#include "gtest/gtest.h"

using namespace frc;

class PIDToleranceTest : public testing::Test {
 protected:
  const double setpoint = 50.0;
  const double range = 200;
  const double tolerance = 10.0;

  class FakeInput {
   public:
    double val = 0;

    double Get() { return val; }
  };

  class FakeOutput {
   public:
    void Set(double output) {}
  };

  FakeInput inp;
  FakeOutput out;
  frc2::PIDController* pidController;

  void SetUp() override {
    pidController = new frc2::PIDController(0.5, 0.0, 0.0);
    pidController->SetInputRange(-range / 2, range / 2);
  }

  void TearDown() override { delete pidController; }

  void Reset() { inp.val = 0; }
};

TEST_F(PIDToleranceTest, Initial) { EXPECT_FALSE(pidController->AtSetpoint()); }

TEST_F(PIDToleranceTest, Absolute) {
  Reset();

  pidController->SetAbsoluteTolerance(tolerance);
  pidController->SetSetpoint(setpoint);

  EXPECT_FALSE(pidController->AtSetpoint())
      << "Error was in tolerance when it should not have been. Error was "
      << pidController->GetError();

  inp.val = setpoint + tolerance / 2;
  for (int i = 0; i <= 50; i++) {
    pidController->Calculate(inp.Get());
  }

  EXPECT_TRUE(pidController->AtSetpoint())
      << "Error was not in tolerance when it should have been. Error was "
      << pidController->GetError();

  inp.val = setpoint + 10 * tolerance;
  for (int i = 0; i <= 50; i++) {
    pidController->Calculate(inp.Get());
  }

  EXPECT_FALSE(pidController->AtSetpoint())
      << "Error was in tolerance when it should not have been. Error was "
      << pidController->GetError();
}

TEST_F(PIDToleranceTest, Percent) {
  Reset();

  pidController->SetPercentTolerance(tolerance);
  pidController->SetSetpoint(setpoint);

  EXPECT_FALSE(pidController->AtSetpoint())
      << "Error was in tolerance when it should not have been. Error was "
      << pidController->GetError();

  inp.val =
      setpoint + (tolerance) / 200 *
                     range;  // half of percent tolerance away from setpoint
  for (int i = 0; i <= 50; i++) {
    pidController->Calculate(inp.Get());
  }

  EXPECT_TRUE(pidController->AtSetpoint())
      << "Error was not in tolerance when it should have been. Error was "
      << pidController->GetError();

  inp.val =
      setpoint +
      (tolerance) / 50 * range;  // double percent tolerance away from setPoint

  for (int i = 0; i <= 50; i++) {
    pidController->Calculate(inp.Get());
  }

  EXPECT_FALSE(pidController->AtSetpoint())
      << "Error was in tolerance when it should not have been. Error was "
      << pidController->GetError();
}
